File: 6.Software-Design-Patterns/example/Creational/Multiton/Multiton.go
package main

import (
	"fmt"
	"sync"
)

// Multiton represents the Multiton pattern
type Multiton struct {
	id string
}

// instances stores instances of Multiton
var (
	instances = make(map[string]*Multiton)
	lock      sync.Mutex
)

// getInstance returns an instance of Multiton associated with the provided key
func getInstance(key string) *Multiton {
	lock.Lock()
	defer lock.Unlock()

	if instances[key] == nil {
		instances[key] = &Multiton{id: key}
	}
	return instances[key]
}

// log prints a log message
func (m *Multiton) log(message string) {
	fmt.Printf("Logger %s: %s\n", m.id, message)
}

func main() {
	example1 := getInstance("app")
	example1.log("This is the first logger instance")

	example2 := getInstance("api")
	example2.log("This is the second logger instance")

	fmt.Println(example1 == example2) // Should print false
}

File: 6.Software-Design-Patterns/example/Creational/Multiton/Multiton.py
class Multiton:
    _instances = {}

    def __new__(cls, key):
        if key not in cls._instances:
            cls._instances[key] = super().__new__(cls)
        return cls._instances[key]

    def __init__(self, key):
        self.id = key

    def log(self, message):
        print(f"Logger {self.id}: {message}")


# Example usage
example_1 = Multiton("app")
example_1.log("This is the first logger instance")

example_2 = Multiton("api")
example_2.log("This is the second logger instance")

print(example_1 is example_2)  # Output: False
